hare/algorithm.go

```diff
@@ -85,7 +87,7 @@ var _ TerminationOutput = (*procReport)(nil)
 
 // State holds the current state of the consensus process (aka the participant).
 type State struct {
-	k           uint32       // the round counter (k%4 is the round number)
+	k           uint32       // the round counter (k%4 is the round number); it should be first in struct for alignment because atomics are used
```

Comment:
it is better to understand when k can be updated and read concurrently. the race here might be the symptom of the real issue

Reply:
IIRC, k was concurrently updated in the test, therefore data races occurred

Reply:
1. 
```
Write at 0x00c000120318 by goroutine 25:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).addToK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:927 +0x84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).advanceToNextRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:545 +0x58
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:337 +0xf66
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:253 +0x58

Previous read at 0x00c000120318 by goroutine 26:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).getK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:917 +0x479
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).sendMessage()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:503 +0x4e0
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:300 +0x284

```

2. 
```
Write at 0x00c000266318 by goroutine 170:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).addToK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:927 +0x84
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).advanceToNextRound()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:545 +0x58
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:337 +0xf66
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).Start·dwrap·1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:253 +0x58

Previous read at 0x00c000266318 by goroutine 173:
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).getK()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:917 +0x33e
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).shouldParticipate()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:866 +0x3a5
  github.com/spacemeshos/go-spacemesh/hare.(*consensusProcess).eventLoop.func1()
      /Users/nkryuchkov/GolandProjects/go-spacemesh/hare/algorithm.go:292 +0x85
```